src/librustc_resolve/late.rs

```diff
@@ -2038,35 +2038,19 @@ impl<'a, 'b> LateResolutionVisitor<'a, '_> {
         }
 
         for &(trait_name, binding) in traits.as_ref().unwrap().iter() {
-            // Traits have pseudo-modules that can be used to search for the given ident.
-            if let Some(module) = binding.module() {
-                let mut ident = ident;
-                if ident.span.glob_adjust(
-                    module.expansion,
-                    binding.span,
-                ).is_none() {
-                    continue
-                }
-                if self.r.resolve_ident_in_module_unadjusted(
```

Comment:
So, what I don't like here is that we are conflating the strict part (determining the set of traits in scope) with the optimization part (pruning the traits that certainly don't fit).
For trait aliases we must perform the first part as well (we don't do that currently, as you can see below), but for them we cannot perform the second part (or at least let's assume that we cannot).

Reply:
So, traits "from macros" should be rejected without using the `ident`.